spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java

```diff
@@ -0,0 +1,197 @@
+import org.springframework.util.Assert;
+
+/**
+ * The {@link AbstractMessageHandler} implemntation to send messages to STOMP destinations.
+ *
+ * @author Artem Bilan
+ * @since 4.2
+ */
+public class StompMessageHandler extends AbstractMessageHandler implements IntegrationEvaluationContextAware,
+		ApplicationEventPublisherAware, Lifecycle {
+
+	private final StompSessionHandler sessionHandler = new IntegrationOutboundStompSessionHandler();
+
+	private final StompSessionManager stompSessionManager;
+
+	private volatile StompSession stompSession;
+
+	private volatile boolean running;
+
+	private volatile HeaderMapper<StompHeaders> headerMapper = new StompHeaderMapper();
+
+	private Expression destinationExpression;
+
+	private EvaluationContext evaluationContext;
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	public StompMessageHandler(StompSessionManager stompSessionManager) {
+		Assert.notNull(stompSessionManager, "'stompSessionManager' is required.");
+		this.stompSessionManager = stompSessionManager;
+	}
+
+	public void setDestination(String destination) {
+		Assert.hasText(destination, "'destination' must not be empty.");
+		this.destinationExpression = new ValueExpression<String>(destination);
+	}
+
+	public void setDestinationExpression(Expression destinationExpression) {
+		Assert.notNull(destinationExpression, "'destinationExpression' must not be null.");
+		this.destinationExpression = destinationExpression;
+	}
+
+	public void setHeaderMapper(HeaderMapper<StompHeaders> headerMapper) {
+		Assert.notNull(headerMapper, "'headerMapper' must not be null.");
+		this.headerMapper = headerMapper;
+	}
+
+	@Override
+	public void setIntegrationEvaluationContext(EvaluationContext evaluationContext) {
+		this.evaluationContext = evaluationContext;
+	}
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
+
+	@Override
+	protected void handleMessageInternal(final Message<?> message) throws Exception {
+		if (!this.isRunning()) {
+			throw new MessageDeliveryException(message, "The StompMessageHandler [" + getComponentName() +
+					"] hasn't been connected to StompSession. Check the state of [" + this.stompSessionManager + "]");
+		}
+
+		StompHeaders stompHeaders = new StompHeaders();
+		this.headerMapper.fromHeaders(message.getHeaders(), stompHeaders);
+		if (stompHeaders.getDestination() == null) {
+			Assert.state(this.destinationExpression != null, "One of 'destination' or 'destinationExpression' must be" +
+					" provided, if message header doesn't supply 'destination' STOMP header.");
+			String destination = this.destinationExpression.getValue(this.evaluationContext, message, String.class);
+			stompHeaders.setDestination(destination);
+		}
+
+		final StompSession.Receiptable receiptable = this.stompSession.send(stompHeaders, message.getPayload());
```

Comment:
Need to wait for the `stompSession` to be established by the callback; maybe a latch set up by `start()` ?